src/c++/perf_analyzer/triton_metrics_manager.h

```diff
@@ -0,0 +1,105 @@
+// Copyright 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <chrono>
+#include <cstdint>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <utility>
+#include <vector>
+#include "client_backend/client_backend.h"
+#include "triton_metrics.h"
+
+namespace triton { namespace perfanalyzer {
+
+// FIXME: pull all this stuff into implementation file :(
+// FIXME: add testing for methods here
+class TritonMetricsManager {
+ public:
+  TritonMetricsManager(
+      std::shared_ptr<clientbackend::ClientBackend> client_backend,
+      uint64_t triton_metrics_interval_ms)
+      : client_backend_(client_backend),
+        triton_metrics_interval_ms_(triton_metrics_interval_ms)
+  {
+  }
+
+  ~TritonMetricsManager() { StopQueryingTritonMetrics(); }
+
+  void StartQueryingTritonMetrics()
+  {
+    should_keep_querying_ = true;
+    query_triton_metrics_main_loop_future_ = std::async(
+        &TritonMetricsManager::QueryTritonMetricsEveryNMilliseconds, this);
+  }
+
+  void QueryTritonMetricsEveryNMilliseconds()
+  {
+    while (should_keep_querying_) {
+      std::future<void> f{
+          std::async(&TritonMetricsManager::QueryTritonMetrics, this)};
```

Comment:
I think we shouldn't start a new thread here using `std::async`. We should just call the API and wait for the results.

Reply:
Discussed offline and I'll remove this lower-level async-ness and just call the metric collection synchronously and wait the remainder of the time 👍

Reply:
Removed lower-level multi-threading 👍